### PR TITLE
Namespacing and global leakage

### PR DESCRIPTION
**Pull request by hillerstorm**
Would at least be a start towards closing #2 :)

Wrapped everything up and exposed only the shared functions and variables in a global object on `window`.

### CODE REVIEW

**Comment by a1k0n** [comment]
Wow, this is great! Thanks for cleaning this mess up!

File: xm.js [hunk 1]
```diff
@@ -279,26 +294,24 @@ function RedrawScreen() {
...
 function next_row() {
-  if (cur_pat == -1 || cur_row >= xm.patterns[cur_pat].length) {
-    cur_row = 0;
-    cur_songpos++;
-    if (cur_songpos >= xm.songpats.length)
-      cur_songpos = xm.song_looppos;
-    cur_pat = xm.songpats[cur_songpos];
+  if (player.cur_pat == -1 || player.cur_row >= player.xm.patterns[player.cur_pat].length) {
+    player.cur_row = 0;
+    player.cur_songpos += 1;
```
Comment (a1k0n) [inline xm.js]: Is += 1 preferred over ++ in Javascript style?

File: xmeffects.js [hunk 2]
```diff
@@ -1,8 +1,14 @@
+(function (window) {
+if (!window.XMPlayer) {
+  window.XMPlayer = {};
+}
+var player = window.XMPlayer;
```
Comment (a1k0n) [inline xmeffects.js:5]: Why does this work? If xmeffects.js loads before xm.js, won't `var player` here be divorced from the actual XMPlayer in xm.js?

File: xmeffects.js [hunk 3]
```diff
@@ -287,3 +293,4 @@ var effects_t1 = [  // effect functions on tick 1+
   eff_unimplemented,  // y
   eff_unimplemented   // z
 ];
+})(window || {});
```
Comment (a1k0n) [inline xmeffects.js]: is `|| {}` this for unit tests or something, so it can run without an actual `window`?

File: xm.js [hunk 4]
```diff
@@ -1258,3 +1271,4 @@ window.onload = function() {
   }
   DownloadXM(uri);
 }
+})(window, document);
```
Comment (a1k0n) [inline xm.js]: I prefer newlines at EOF, if possible

File: xm.js (hunk 1 above)
Comment (hillerstorm) [inline xm.js]: Ah, nah I just brainfarted and thought I couldn't ++ fields on objects. But that's not true at all :) I'll fix it!

File: xmeffects.js (hunk 2 above)
Comment (hillerstorm) [inline xmeffects.js:5]: `var player = window.XMPlayer` is just by reference. Any changes to `window.XMPlayer` will reflect on `player`, I just didn't want to write the longer name everywhere :)

File: xmeffects.js (hunk 3 above)
Comment (hillerstorm) [inline xmeffects.js]: Yea that's basically what I was thinking. One could swap it out for some other module implementation as a fallback if `window` is undefined (running in a non-browser)

File: xmeffects.js (hunk 2 above)
Comment (a1k0n) [inline xmeffects.js:5]: Oh, I see. Either xmeffects.js or xm.js will create the `window.XMPlayer` and each will use the other's if it already exists. OK.

File: xmeffects.js (hunk 3 above)
Comment (a1k0n) [inline xmeffects.js]: Ok. Let's be consistent between xm.js and xmeffects.js then. I think in a unit test setting we'd just override window anyway, right?

File: xm.js (hunk 4 above)
Comment (hillerstorm) [inline xm.js]: As you wish! :)

File: xmeffects.js (hunk 3 above)
Comment (hillerstorm) [inline xmeffects.js]: Yep, and then there's `document` but anything that uses it has to be tested with some headless browser anyways. Which doesn't seem to be a very high priority in a project like this so we can just leave it as it is.

Adding the fallback to xm.js

Comment (a1k0n) [inline xmeffects.js]: Good point. I actually meant just removing the || {} from window since we'd be overriding it, but this is fine...

**Comment by a1k0n** [comment]
Looks good. If you're happy with it now, I'll merge.

**Comment by hillerstorm** [comment]
:shipit: :+1:

**Comment by a1k0n** [comment]
Thanks again!

**Comment by tibbon** [comment]
Nice work!

**Pull request merged**